### PR TITLE
rgwuadmin pyx

### PR DESCRIPTION
**Pull request by mattbenjamin**
Hi @orit,

I made some assumptions about how user_info was supposed to work, but I think this basically works, as far as it goes?  (I was going to have Ali do this, but decided I wanted to learn how Cython worked.)

Matt

# create a user
uid = "tlhm"
display_name = "thierry lhermite"
access_key = "strawberry"
secret_key = "rhubarb"
email = "thierry@whouse.gov"
caps = ""
access = ""
admin = True
system = False
ret = rgw_adm.create_user(uid, display_name, access_key, secret_key, email, caps, access, admin, system)

>>> ret = rgw_adm.create_user(uid, display_name, access_key, secret_key, email, caps, access, admin, system)
... 
>>> ret = rgw_adm.user_info(uid)
>>> 
>>> ret
UserInfo(uid='tlhm', display_name='thierry lhermite', access_key='strawberry', secret_key='rhubarb', email='thierry@whouse.gov', caps='', access='', admin=True, system=True)
>>>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
Fixed admin and system output.  Still need access and caps.
N.b., init starts  around 70 threads, and seems like 57 exiting on shutdown...

>>> ret = rgw_adm.user_info(uid)

Thread 1 "python" hit Breakpoint 3, rgw_admin_user_info (librgw_admin_user=0x55555588e1a0, 
    uid=uid@entry=0x7ffff7e8f5f4 "bob1", user_info=user_info@entry=0x7fffffffd730)
    at /home/mbenjamin/ceph-noob/src/rgw/rgw_admin_user.cc:107
107	  {
(gdb) c
Continuing.
>>> ret
UserInfo(uid='bob1', display_name='bob1', access_key='bob1.access', secret_key='bob1.key', email='bob1@whouse.gov', caps='', access='', admin=True, system=True)
>>> rgw_adm.shutdown()
2018-04-20 10:19:34.675 7ffff7fcf4c0 -1 shutting down
[Thread 0x7fffdce8f700 (LWP 9436) exited]
[Thread 0x7fffb5fdb700 (LWP 9427) exited]
[Thread 0x7fffc27f4700 (LWP 9402) exited]
[Thread 0x7fffbefed700 (LWP 9409) exited]
...
57 more threads
...
>>>

**Comment by LenzGr** [comment]
@p-na, @votdev or @sebastian-philipp - can you please take a look if and how this can be used from the dashboard backend to create the RGW admin user? Thanks!

**Comment by sebastian-philipp** [comment]
How does this relate to #21439 ?

**Comment by p-na** [comment]
It seems that the dashboard mgr module would benefit from these changes. Using those changes it would be capable of automatically creating a dedicated RGW admin user to provide the RGW management functionality, reducing the required manual steps of the user/administrator to configure the RGW management functionality. No need to create the user and configure the dashboard with its credentials anymore.

Hence, we're looking forward for this PR to be merged, but it's been a while since the latest update. Any news on this one?

**Comment by p-na** [comment]
This is a nice feature for the dashboard! May I ask for an update? Is this going to be implemented?

**Comment by mattbenjamin** [comment]
(rebasing)

**Comment by epuertat** [comment]
As mentioned @mattbenjamin, maybe replacing `rgw` with `rgw/file` and `rgw/admin` in [this file](../blob/master/admin/build-doc#L79) would fix the doc build? I don't have write access to Orit's repo

**Comment by mattbenjamin** [comment]
@epuertat I'll give it a try (it's my repo :)

**Comment by mattbenjamin** [comment]
@LenzGr @epuertat could I get a review, or a recommendation for a review?

thanks,

Matt

**Comment by epuertat** [comment]
@mattbenjamin seems the doc build is still failing (now for a different reason). We may count on someone with more knowledge on this (@tchaikov? @joscollin?). Thanks!

**Comment by mattbenjamin** [comment]
@epuertat the doc build succeeded, that's why I pinged you;   the make check failure is unrelated to this change

**Comment by epuertat** [comment]
retest this please

**Comment by epuertat** [comment]
@mattbenjamin thanks! I just saw a failing check and thought it was the same as yesterday. This one looks weird (how can the objectstore can be broken by rgw)? I just relaunched tests.

**Comment by LenzGr** [comment]
Adding some potential reviewers from the dashboard team to review/comment.

**Comment by mattbenjamin** [comment]
@votdev ^^

**Comment by LenzGr** [comment]
Adding the dashboard label, so it's on our radar and can be reviewed (as the dashboard would be a first candidate to consume this API)

File: src/pybind/rgw_admin_user/rgw_admin_user.pyx [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    # get rgw user info
+    int rgw_admin_user_info(librgw_admin_user_t rgw_admin_user, const_char *uid,
+    rgw_user_info *user_info)
+
+    # free user_info structure
+    void rgw_admin_user_release_info(rgw_user_info* user_info)
+
+UserInfo = namedtuple('UserInfo', ["uid", "display_name", "access_key", "secret_key", "email", "caps", "access", "admin", "system"])
+    
+class Error(Exception):
```
Comment (sebastian-philipp) [inline src/pybind/rgw_admin_user/rgw_admin_user.pyx]: Just curious: why do we need to copy the exceptions all over again? As the exceptions are everywhere completely identical, would it make sense to just have them at one location?

Comment (tchaikov) [reply src/pybind/rgw_admin_user/rgw_admin_user.pyx]: @sebastian-philipp i think the reason is that we still don't have a package which is shared across different python bindings. we don't have a package listed as a dependency of all python binding packages. i think the mgr modules also suffer from the same problem -- how to share/reuse the common bits.

Comment (sebastian-philipp) [reply src/pybind/rgw_admin_user/rgw_admin_user.pyx]: The mgr modules can share common code in the pybind/mgr dir, but there is indeed no real solution for code common between ceph components.

Comment (tchaikov) [reply src/pybind/rgw_admin_user/rgw_admin_user.pyx]: ahh, right! i just realized that we have extracted `mgr_util.py`. IMHO, `mgr_module.py` is kinda awkward as we are dumping more and more utilities into it.

**Comment by sebastian-philipp** [comment]
what's missing to get this PR merged?

**Comment by p-na** [comment]
@votdev Can you please review this PR?

File: src/pybind/rgw_admin_user/rgw_admin_user.pyx [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        uid = cstr(uid, 'uid')
+
+        cdef:
+            const_char *_uid = uid
+            rgw_user_info user_info
+
+        with nogil:
+            ret = rgw_admin_user_info(self.handle, _uid, &user_info)
+
+        info = UserInfo(uid=user_info.uid,
```
Comment (votdev) [inline src/pybind/rgw_admin_user/rgw_admin_user.pyx]: I'm missing the ``tenant`` field here. Working on the RGW part in the Dashboard showed me that this field is really important when you work with tenanted users and RGW Admin OPS API. The Admin OPS API is forcing you to generate a ``uid`` which contains the tenant name, see https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/controllers/rgw.py#L109 and  https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/controllers/rgw.py#L156.

File: src/rgw/rgw_admin_user.cc [hunk 3]
```diff
@@ -62,7 +74,42 @@ extern "C" {
...
+  /* transfer object */
+  struct RGWLA_User
+  {
+    RGWUserInfo info;
+    std::string uid;
+    std::string display_name;
+    std::string access_key;
+    std::string secret_key;
+    std::string caps;
+    std::string access;
```
Comment (votdev) [inline src/rgw/rgw_admin_user.cc]: The ``tenant`` name is missing.

File: src/test/rgw/test_user_admin.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    process = subprocess.Popen(cmd, **kwargs)
+    s = process.communicate()[0]
+    log.debug('command returned status=%d stdout=%s', process.returncode, s.decode('utf-8'))
+    if check_retcode:
+        assert(process.returncode == 0)
+    return (s, process.returncode)
+
+
+class TestUserAdmin:
+    """test driver"""
```
Comment (votdev) [inline src/test/rgw/test_user_admin.py]: Please add some tests with tenanted users.

**Comment by rjfd** [comment]
@votdev @LenzGr  are we still planning to improve RGW management by creating automatically an admin user with this API?

File: src/pybind/rgw_admin_user/rgw_admin_user.pyx (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/rgw_admin_user/rgw_admin_user.pyx]: See #28837 a solution to this:
```suggestion
from ceph import exception
```

**Comment by LenzGr** [comment]
After some further discussions we (the Dashboard team) concluded that we don't require this functionality for the time being. We still think it would be a nice enhancement of the Python bindings, though (they should be on par with the C++-library). Removing the "dashboard" label for now.

File: src/pybind/rgw_admin_user/rgw_admin_user.pyx (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/rgw_admin_user/rgw_admin_user.pyx]: yeah, this should now start to work. Might need to make `python-ceph-common` available in this context, though.

**Comment by cbodley** [comment]
> After some further discussions we (the Dashboard team) concluded that we don't require this functionality for the time being. We still think it would be a nice enhancement of the Python bindings, though (they should be on par with the C++-library). Removing the "dashboard" label for now.

thanks @LenzGr! closing

**Pull request closed without merging**